x-pack/legacy/plugins/reporting/server/legacy.ts

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import { Legacy } from 'kibana';
+import { PluginInitializerContext } from 'src/core/server';
+import { PluginSetupContract as SecurityPluginSetup } from '../../../../plugins/security/server';
+import { ReportingPluginSpecOptions } from '../types';
+import { plugin } from './index';
+import { LegacySetup, ReportingStartDeps } from './plugin';
+
+const buildLegacyDependencies = (
+  server: Legacy.Server,
+  reportingPlugin: ReportingPluginSpecOptions
+): LegacySetup => ({
+  config: server.config,
+  info: server.info,
+  route: server.route.bind(server),
+  plugins: {
+    elasticsearch: server.plugins.elasticsearch,
+    xpack_main: server.plugins.xpack_main,
+    reporting: reportingPlugin,
+  },
+  savedObjects: server.savedObjects,
+  uiSettingsServiceFactory: server.uiSettingsServiceFactory,
+});
+
+export const legacyInit = async (
+  server: Legacy.Server,
+  reportingPlugin: ReportingPluginSpecOptions
+) => {
+  const coreSetup = server.newPlatform.setup.core;
+  const pluginInstance = plugin(server.newPlatform.coreContext as PluginInitializerContext);
+
+  await pluginInstance.setup(coreSetup, {
```

Comment:
Do you have to await the result of setup?
The next stage awaits on the results of getStartServices anyway.

Reply:
I do need a way to make sure that my plugin's `start` method would not be called before `setup`. 

Soon in the plugin class there will be a private field referenced in `setup` and `start`, so I'd have a race condition if `start` was using the data before it is initialized in `setup`.

I don't have my mind settled yet on how to co-ordinate dependencies in setup and start, but I plan to have it outlined in an upcoming PR to provide more plugin dependencies to modules that aren't called from routes handlers. :)